### PR TITLE
Homework2

### PR DESCRIPTION
**Pull request by Crabs-hub**
Homework2

### CODE REVIEW

File: src/main/java/Problem1.java [hunk 1]
```diff
@@ -1,10 +1,34 @@
...
+            if(nums[i] == 0){
+                temp = nums[i];
+                nums[i] = nums[left];
+                nums[left] = temp;
+                left++;
+            }
+            else if(nums[i] == 2){
+                temp = nums[i];
+                nums[i] = nums[right - 1];
+                nums[right - 1] = temp;
```
Comment (pdgetrf) [inline src/main/java/Problem1.java:28]: I see duplication here. consider refactoring into a private swap function.

we prefer to have "single source of truth" in our code.

File: src/main/java/Problem2.java [hunk 2]
```diff
@@ -2,11 +2,27 @@
...
+        if (nums.length == 1){
+            return unique;
+        }
+        for(int i = 0; i < nums.length - 1; i++){
+            if(nums[i] != nums[i+1]){
+                nums[unique] = nums[i+1];
+                unique++;
+            }
+        }
+        return unique; // place holder
```
Comment (pdgetrf) [inline src/main/java/Problem2.java:26]: remove the "place holder" comment

File: src/main/java/Problem3/Book.java [hunk 3]
```diff
@@ -15,11 +15,18 @@ public Book(String title, String author) {
...
+        this.id = anotherBook.id;
+        this.title = anotherBook.title;
+        this.author = anotherBook.author;
     }
 
     @Override
     public boolean equals(Object obj) {
         // homework
+        Book anotherBook = (Book) obj;
+        boolean equivalent = (this.id == anotherBook.id) && (this.title == anotherBook.title) &&
```
Comment (pdgetrf) [inline src/main/java/Problem3/Book.java:28]: wrong logic 

```
A movie equals another movie only when their id are the same. Same rule for books being equal.
```

File: src/main/java/Problem3/Movie.java [hunk 4]
```diff
@@ -10,14 +10,23 @@
...
+        this.rating = anotherMovie.rating;
+        this.title = anotherMovie.title;
+        this.id = anotherMovie.id;
     }
 
     @Override
     public boolean equals(Object obj) {
         // homework
+        Movie anotherMovie = (Movie) obj;
+        boolean equivalence = (this.id == anotherMovie.id) && (this.rating == anotherMovie.rating) &&
```
Comment (pdgetrf) [inline src/main/java/Problem3/Movie.java:28]: same here

File: src/main/java/Problem3/MovieAction.java [hunk 5]
```diff
@@ -22,5 +24,12 @@ public int getLateFeeInDollar() {
     @Override
     public int calcLateFees(int numOfDaysPastDue) {
         // homework
+        if(numOfDaysPastDue < 0){
+            return 0;
+        }
+        else if(numOfDaysPastDue < 5){
+            return numOfDaysPastDue * lateFeePerDayInDollar;
+        }
+        return 2 * numOfDaysPastDue * lateFeePerDayInDollar;
```
Comment (pdgetrf) [inline src/main/java/Problem3/MovieAction.java:33]: also repeated logic here. suggesting:

```java
        int fee = lateFeePerDayInDollar * numOfDaysPastDue;
        return (numOfDaysPastDue >= 5) ? 2 * fee : fee;
```

File: src/test/java/Problem2Test.java [hunk 6]
```diff
@@ -7,13 +7,23 @@
...
 
         int[][] expects = {
-                {0, 1, 2, 4, 6}
+                {0, 1, 2, 4, 6},
                 // add more test cases
+                {0, 1, 2, 3, 5, 6},
+                {0, 1, 2, 3, 4, 6},
+                {0, 1, 2, 3, 4, 5},
+                {1},
+                {2},
```
Comment (pdgetrf) [inline src/test/java/Problem2Test.java:26]: great test. 

add a few more for edge cases such as null array and empty array.

**Review by pdgetrf — APPROVED** [review]
LGTM
